controller/controller_server/utils.py

```diff
@@ -8,43 +10,164 @@
+            raise ValidationException(messages.invalid_secrets_message)
+    return system_info, system_id
+
+
+def _get_array_connection_info_from_system_info(secrets, system_id):
     user = secrets[config.SECRET_USERNAME_PARAMETER]
     password = secrets[config.SECRET_PASSWORD_PARAMETER]
-    array_addresses = secrets[config.SECRET_ARRAY_PARAMETER].split(",")
-    return user, password, array_addresses
+    array_addresses = secrets[config.SECRET_ARRAY_PARAMETER].split(config.PARAMETERS_ARRAY_ADDRESSES_DELIMITER)
+    return ArrayConnectionInfo(array_addresses=array_addresses, user=user, password=password, system_id=system_id)
+
+
+def get_array_connection_info_from_secrets(secrets, topologies=None, system_id=None):
+    system_info, system_id = _get_system_info_from_secrets(secrets, topologies, system_id)
+    return _get_array_connection_info_from_system_info(system_info, system_id)
+
+
+def get_volume_parameters(parameters, system_id):
+    return get_object_parameters(parameters, config.PARAMETERS_VOLUME_NAME_PREFIX, system_id)
+
 
+def get_snapshot_parameters(parameters, system_id):
+    return get_object_parameters(parameters, config.PARAMETERS_SNAPSHOT_NAME_PREFIX, system_id)
 
-def get_volume_id(new_volume):
-    return _get_object_id(new_volume)
+
+def get_object_parameters(parameters, prefix_param_name, system_id):
+    raw_parameters_by_system = parameters.get(config.PARAMETERS_BY_SYSTEM)
+    system_parameters = {}
+    if raw_parameters_by_system and system_id:
+        parameters_by_system = _parse_raw_input(raw_input=raw_parameters_by_system)
+        system_parameters = parameters_by_system.get(system_id, {})
+    default_pool = parameters.get(config.PARAMETERS_POOL)
+    default_space_efficiency = parameters.get(config.PARAMETERS_SPACE_EFFICIENCY)
+    default_prefix = parameters.get(prefix_param_name)
+    return ObjectParameters(
+        pool=system_parameters.get(config.PARAMETERS_POOL, default_pool),
+        space_efficiency=system_parameters.get(config.PARAMETERS_SPACE_EFFICIENCY, default_space_efficiency),
+        prefix=system_parameters.get(prefix_param_name, default_prefix))
+
+
+def get_volume_id(new_volume, system_id):
+    return _get_object_id(new_volume, system_id)
 
 
 def get_snapshot_id(new_snapshot):
     return _get_object_id(new_snapshot)
 
 
-def _get_object_id(obj):
+def _get_object_id(obj, system_id=None):
+    if system_id:
+        return config.PARAMETERS_OBJECT_ID_DELIMITER.join((obj.array_type, system_id, obj.id))
     return config.PARAMETERS_OBJECT_ID_DELIMITER.join((obj.array_type, obj.id))
 
 
-def validate_secret(secret):
-    logger.debug("validating secrets")
-    if secret:
-        if not (config.SECRET_USERNAME_PARAMETER in secret and
-                config.SECRET_PASSWORD_PARAMETER in secret and
-                config.SECRET_ARRAY_PARAMETER in secret):
-            raise ValidationException(messages.invalid_secret_message)
+def _is_imported_string_valid(imported_string):
```

Comment:
rename to something more specific, e.g. `_is_system_id_valid`/`_is_valid_system_id`

Reply:
done